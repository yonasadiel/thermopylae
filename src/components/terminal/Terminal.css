.terminal {
  width: 50vw;
  margin: 10px auto;
  position: relative;
}

.terminal input {
  width: 100%;
  background: transparent;
  color: var(--foreground-color);
  border-width: 0 0 2px 0;
  padding: 5px 0;
  outline: none!important;
  font-family: 'Raleway', sans-serif;
  font-size: 24px;
  text-align: left;
}

#bangs {
  position: absolute;
  width: 100%;
  background-color: rgba(200, 200, 200, .2);
  z-index: 100;
}

#bangs .suggestion {
  padding: 0.5em 0 0.5em 0;
  border: 0 solid var(--foreground-color);
  border-width: 0 0 0px 0;
}

#bangs .suggestion .pattern {
  font-size: 24px;
  color: var(--foreground-color);
  opacity: .5;
  margin-bottom: 0.25em;
}

#bangs .suggestion.selected {
  background-color: rgba(255, 255, 255, 0.1);
}

#bangs .suggestion.selected .pattern {
  opacity: 1;
  font-weight: bold;
}
