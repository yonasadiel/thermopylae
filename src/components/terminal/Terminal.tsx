import { KeyboardEvent, act, useEffect, useMemo, useRef, useState } from 'react';
import { filterBangs, loadBangs, preprocessBangs, processBang, recordHistory } from './util';
import { Bang, ProcessedBang } from '../../models/terminal';
import './Terminal.css';

export default function Terminal() {
    const inputRef = useRef<HTMLInputElement>(null);
    const [query, setQuery] = useState<string>('');
    const [loadedBangs, setLoadedBangs] = useState<Bang[]>([]);
    const filteredBangs = useMemo(() => {
        const firstWord = query.split(' ')[0] || '';
        return filterBangs(loadedBangs, firstWord);
    }, [query, loadedBangs]);
    const processedBang = useMemo<ProcessedBang | null>(() => {
        if (!!filteredBangs && filteredBangs.length > 0) return processBang(filteredBangs[0], query);
        return null;
    }, [filteredBangs, query]);
    useEffect(() => { inputRef.current?.focus(); }, [inputRef]);
    useEffect(() => {
        loadBangs().then((res) => {
            const newLoadedBangs = preprocessBangs(res);
            act(() => setLoadedBangs(newLoadedBangs));
            console.log(`Loaded ${newLoadedBangs.length} bangs`);
        });
    }, []);

    const handleKeyDown = (ev?: KeyboardEvent<HTMLInputElement>) => {
        if (ev?.code === 'Enter') {
            if (!!processedBang) {
                recordHistory(processedBang.history);
                window.location.replace(processedBang.target);
            } else {
                console.warn('no bang available'); // TODO: show toast or something
            }
        } else if (ev?.code === 'Tab') {
            ev.preventDefault();
            if (!!processedBang && !!processedBang.suggestions) {
                setQuery(processedBang.suggestions[0]);
            } else {
                console.warn('no suggestion available'); // TODO: show toast or something
            }
        }
    };

    return (
        <div className="terminal">
            <input
                type="text"
                value={query}
                onChange={(e) => setQuery(e.currentTarget.value)}
                onKeyDown={(ev) => handleKeyDown(ev)}
                ref={inputRef}
                list="bangs" />
            <div
                id="bangs">
                {!!processedBang && (
                    <>
                        <div className="target">{processedBang.target}</div>
                        {processedBang.suggestions.map((suggestion) => (
                            <div className="suggestion" key={suggestion}>
                                <p className="pattern">{suggestion}</p>
                            </div>
                        ))}
                    </>
                )}
            </div>
        </div>
    );
};

