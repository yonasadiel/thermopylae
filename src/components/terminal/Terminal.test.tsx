import { test, expect, describe } from '@jest/globals';
import { render } from '@testing-library/react';
import Terminal from './Terminal';

describe('Terminal', () => {
    test('should render correctly', () => {
        const result = render(<Terminal />);
        expect(result.baseElement).toMatchSnapshot();
    });
});