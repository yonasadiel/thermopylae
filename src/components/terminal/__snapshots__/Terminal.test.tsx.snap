// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Terminal should render correctly 1`] = `
<body>
  <div>
    <div
      class="terminal"
    >
      <input
        list="bangs"
        type="text"
        value=""
      />
      <div
        id="bangs"
      />
    </div>
  </div>
</body>
`;
